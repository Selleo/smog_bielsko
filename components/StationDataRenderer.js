import React, { Component, PropTypes } from 'react';
import { View, StyleSheet, Text } from 'react-native';

import Env from '../env.js'
import CityInfo from './city/CityInfo'
import PollutionValues from './pollution/PollutionValues'
import AirQualityIndex from './pollution/AirQualityIndex'

export default class StationDataRenderer extends Component {
  static get defaultProps() {
    return {
      stationId: 6535
    };
  }

  constructor(props) {
    super(props);
    this.state = {
      data: undefined,
      pending: true
    };
  }

  componentWillMount() {
    this.getData();
  }

  render() {
    if (this.state.pending) {
      return (
        <View style={styles.loadingContainer}>
          <Text style={styles.loadingText}>Loading</Text>
        </View>
      )
    } else {
      return (
        <View>
          <CityInfo city={this.state.data.city} />
          <AirQualityIndex index={this.state.data.aqi} />
          <PollutionValues dataset={this.state.data.iaqi} />
        </View>
      )
    }
  }

  getData() {
    return fetch('http://api.waqi.info/feed/@' + this.props.stationId + '/?token=' + Env.API_KEY)
      .then((response) => response.json())
      .then((responseJson) => {
        return this.setState({data: responseJson.data, pending: false})
      })
  }

  static propTypes = {
    stationId: PropTypes.number.isRequired,
  };
}

const styles = StyleSheet.create({
  loadingContainer: {
    flex: 1,
    flexDirection: 'column',
    justifyContent: 'space-around'
  },
  loadingText: {
    fontSize: 18,
    textAlign: 'center'
  }
});
