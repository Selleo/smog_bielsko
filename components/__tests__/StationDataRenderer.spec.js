import 'react-native';
import React                          from 'react';

import { shallow }                    from 'enzyme';
import chai, { expect }               from 'chai';
import dirtyChai                      from 'dirty-chai';
import fetchMock                      from 'fetch-mock'

import StationDataRenderer            from '../StationDataRenderer';
import Env                            from '../../env'

chai.use(dirtyChai);

let wrapper, props;

describe('StationDataRenderer', () => {
  it('renders correctly', () => {
    props = { stationId: 1 };
    fetchMock.get('http://api.waqi.info/feed/@1/?token=' + Env.API_KEY, {
      data: {
        city: { name: 'Zywiec, ul. Rynek 2' },
        aqi: 2,
        iaqi: 2
      }
    });
    wrapper = shallow(<StationDataRenderer {...props}/>);

    expect(wrapper).to.exist();
    fetchMock.restore();
  });

  it('set in state fetched data', (done) => {
    props = { stationId: 1 };
    fetchMock.get('http://api.waqi.info/feed/@1/?token=' + Env.API_KEY, {
      data: {
        city: { name: 'Zywiec, ul. Rynek 2' },
        aqi: 2,
        iaqi: 2
      }
    });
    wrapper = shallow(<StationDataRenderer {...props}/>);

    expect(wrapper.state().data).to.be.an('undefined');
    expect(wrapper.state().pending).to.equal(true);

    setTimeout(() => {
      expect(wrapper.state().data).to.be.an('object');
      expect(wrapper.state().data).to.eql({ city: { name: 'Zywiec, ul. Rynek 2' }, aqi: 2, iaqi: 2 });
      done();
    }, 1);

    fetchMock.restore();
  });

  describe('pending is false', () => {
    it('is false after download data', (done) => {
      props = { stationId: 1 };
      fetchMock.get('http://api.waqi.info/feed/@1/?token=' + Env.API_KEY, {
        data: {
          city: { name: 'Zywiec, ul. Rynek 2' },
          aqi: 2,
          iaqi: 2
        }
      });
      wrapper = shallow(<StationDataRenderer {...props}/>);

      expect(wrapper.state().pending).to.equal(true);

      setTimeout(() => {
        expect(wrapper.state().pending).to.equal(false);
        done();
      }, 1);

      fetchMock.restore();
    });

    it('return text "Loading"', (done) => {
      props = { stationId: 1 };
      fetchMock.get('http://api.waqi.info/feed/@1/?token=' + Env.API_KEY, {
        data: {
          city: { name: 'Zywiec, ul. Rynek 2' },
          aqi: 2,
          iaqi: 2
        }
      });
      wrapper = shallow(<StationDataRenderer {...props}/>);

      expect(wrapper.state().pending).to.equal(true);
      expect(wrapper.children().length).to.equal(1);

      setTimeout(() => {
        expect(wrapper.state().pending).to.equal(false);
        expect(wrapper.children().children().text()).to.equal('Loading');
        expect(wrapper.children().type().name).to.equal('Component');
        done();
      }, 1);

      fetchMock.restore();
    });
  });

  describe('pending is true', () => {
    it('return text "Loading"', (done) => {

      props = { stationId: 1 };
      fetchMock.get('http://api.waqi.info/feed/@1/?token=' + Env.API_KEY, {
        data: {
          city: { name: 'Zywiec, ul. Rynek 2' },
          aqi: 2,
          iaqi: 2
        }
      });
      wrapper = shallow(<StationDataRenderer {...props}/>);

      expect(wrapper.state().pending).to.equal(true);
      expect(wrapper.children().length).to.equal(1);

      setTimeout(() => {
        expect(wrapper.state().pending).to.equal(false);
        wrapper.update();

        expect(wrapper.children().nodes.length).to.equal(3);
        expect(wrapper.children().nodes[0].type.name).to.equal('CityInfo');
        expect(wrapper.children().nodes[1].type.name).to.equal('AirQualityIndex');
        expect(wrapper.children().nodes[2].type.name).to.equal('PollutionValues');

        done();
      }, 1);

      fetchMock.restore();
    });
  });


});

