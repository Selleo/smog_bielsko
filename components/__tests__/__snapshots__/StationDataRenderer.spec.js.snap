exports[`StationDataRenderer renders correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <View
    style={
        Object {
            "flex": 1,
            "flexDirection": "column",
            "justifyContent": "space-around",
          }
    }>
    <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
            Object {
                "fontSize": 18,
                "textAlign": "center",
              }
        }>
        Loading
    </Text>
</View>,
  "nodes": Array [
    <View
      style={
            Object {
                  "flex": 1,
                  "flexDirection": "column",
                  "justifyContent": "space-around",
                }
      }>
      <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
                  Object {
                        "fontSize": 18,
                        "textAlign": "center",
                      }
            }>
            Loading
      </Text>
</View>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <StationDataRenderer
        stationId={1} />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StationDataRenderer {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "stationId": 1,
        },
        "refs": Object {},
        "state": Object {
          "data": undefined,
          "pending": true,
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <View
          style={
                    Object {
                              "flex": 1,
                              "flexDirection": "column",
                              "justifyContent": "space-around",
                            }
          }>
          <Text
                    accessible={true}
                    allowFontScaling={true}
                    ellipsizeMode="tail"
                    style={
                              Object {
                                        "fontSize": 18,
                                        "textAlign": "center",
                                      }
                    }>
                    Loading
          </Text>
</View>,
        "_debugID": 2,
        "_renderedOutput": <View
          style={
                    Object {
                              "flex": 1,
                              "flexDirection": "column",
                              "justifyContent": "space-around",
                            }
          }>
          <Text
                    accessible={true}
                    allowFontScaling={true}
                    ellipsizeMode="tail"
                    style={
                              Object {
                                        "fontSize": 18,
                                        "textAlign": "center",
                                      }
                    }>
                    Loading
          </Text>
</View>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <StationDataRenderer
    stationId={1} />,
}
`;
